apps/cowswap-frontend/src/modules/usdAmount/apis/getDefillamaUsdPrice.ts

```diff
@@ -0,0 +1,73 @@
+import { FractionUtils } from '@cowprotocol/common-utils'
+import { SupportedChainId } from '@cowprotocol/cow-sdk'
+import { Fraction, Token } from '@uniswap/sdk-core'
+
+import ms from 'ms.macro'
+
+import { fetchWithRateLimit } from 'common/utils/fetch'
+
+import { RateLimitError, UnknownCurrencyError } from './errors'
+
+interface DefillamaUsdQuote {
+  coins: {
+    [chainAndAddress: string]: {
+      price: number
+    }
+  }
+}
+
+export const DEFILLAMA_PLATFORMS: Record<SupportedChainId, string | null> = {
+  [SupportedChainId.MAINNET]: 'ethereum',
+  [SupportedChainId.GNOSIS_CHAIN]: 'xdai',
+  [SupportedChainId.SEPOLIA]: null,
+}
+
+const BASE_URL = 'https://coins.llama.fi/prices/current'
+/**
+ * This is a text of 429 HTTP code
+ * https://saturncloud.io/blog/catching-javascript-fetch-failing-with-cloudflare-429-missing-cors-header/
+ */
+const FAILED_FETCH_ERROR = 'Failed to fetch'
+
+const fetchRateLimited = fetchWithRateLimit({
+  // Allow 2 requests per second
+  rateLimit: {
+    tokensPerInterval: 2,
+    interval: 'second',
+  },
+  // 2 retry attempts with 100ms delay
+  backoff: {
+    maxDelay: ms`0.1s`,
+    numOfAttempts: 2,
+  },
+})
+
+export const DEFILLAMA_RATE_LIMIT_TIMEOUT = ms`1m`
+
+export async function getDefillamaUsdPrice(currency: Token): Promise<Fraction | null> {
+  const platform = DEFILLAMA_PLATFORMS[currency.chainId as SupportedChainId]
+
+  if (!platform) throw new Error('UnsupportedDefillamaPlatformError')
+
+  const key = `${platform}:${currency.address}`.toLowerCase()
+  const url = `${BASE_URL}/${key}`
+
+  return fetchRateLimited(url)
+    .then((res) => res.json())
+    .catch((error) => {
+      if (error.message.includes(FAILED_FETCH_ERROR)) {
+        throw new RateLimitError()
```

Comment:
is this right? FAILED_FETCH_ERROR necesarily mean RateLimitError? 

Also, lets not get the cause lost

Reply:
Copied from what's used in coingecko https://github.com/cowprotocol/cowswap/blob/feat/4238_defillama/apps/cowswap-frontend/src/modules/usdAmount/apis/getCoingeckoUsdPrice.ts#L26-L30

To be honest, I wasn't able to get a 429 from defillama while testing, not sure this applies.